server/src/com/thoughtworks/go/server/service/plugins/builder/PluginInfoBuilder.java

```
@@ -0,0 +1,74 @@
+import com.thoughtworks.go.plugin.access.notification.NotificationPluginRegistry;
+import com.thoughtworks.go.plugin.access.packagematerial.JsonBasedPackageRepositoryExtension;
+import com.thoughtworks.go.plugin.access.pluggabletask.JsonBasedTaskExtension;
+import com.thoughtworks.go.plugin.access.scm.SCMExtension;
+import com.thoughtworks.go.plugin.infra.PluginManager;
+import com.thoughtworks.go.server.ui.plugins.PluginInfo;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@Component
+public class PluginInfoBuilder {
+
+    private Map<String, ViewModelBuilder> pluginExtensionToBuilder = new HashMap<>();
+
+    @Autowired
+    public PluginInfoBuilder(AuthenticationPluginRegistry authenticationPluginRegistry, NotificationPluginRegistry notificationPluginRegistry, PluginManager pluginManager) {
+        pluginExtensionToBuilder.put(AuthenticationExtension.EXTENSION_NAME, new AuthenticationViewModelBuilder(pluginManager, authenticationPluginRegistry));
+        pluginExtensionToBuilder.put(NotificationExtension.EXTENSION_NAME, new NotificationViewModelBuilder(pluginManager, notificationPluginRegistry));
+        pluginExtensionToBuilder.put(JsonBasedPackageRepositoryExtension.EXTENSION_NAME, new PackageViewModelBuilder(pluginManager));
+        pluginExtensionToBuilder.put(JsonBasedTaskExtension.TASK_EXTENSION, new PluggableTaskViewModelBuilder(pluginManager));
+        pluginExtensionToBuilder.put(SCMExtension.EXTENSION_NAME, new SCMViewModelBuilder(pluginManager));
+    }
+
+    public List<PluginInfo> allPluginInfos(String type) {
+        List<PluginInfo> plugins = new ArrayList<>();
+
+        if (type != null) {
+            return pluginExtensionToBuilder.containsKey(type) ? pluginExtensionToBuilder.get(type).allPluginInfos() : plugins;
```

If the key is invalid(meaning not a known extension type), could we return an error message instead of returning an empty plugin list?